torch/csrc/autograd/custom_function.cpp

```diff
@@ -339,6 +341,12 @@ optional_variable_list _process_backward_mode_ad(
         impl::rebase_history(var, {cdata, output_nr});
       }
     } else if (is_input) {
+      TORCH_CHECK(
```

Comment:
Is it possible for (!is_differentiable && is_input) to be true? Is this the case where input does not require grad, so it is not ambiguous if we save the input or the input-returned-as-output for backward?

(If yes, then could we add a test for this to check that this case doesn't error out?)

Reply:
After testing with functorch some more, I realized that an error actually isn't actually raised in functorch unless the inner tensors wrapped by GradWrapper themselves require grad. I think its because to the perspective of the outer apply, the function never returned an input as-is (since the inner apply actually applied alias).

One way to fix that is actually to error in the !is_differentiable case as well, so we DO want to error out in this case. I think its reasonable to do this because the whether or not we pass in tensors that require grad, the code is still "bad" (the forward still returns input as-is, and setup_context still saves that tensor for backward), so we should make the user fix their code.

Reply:
> I think its because to the perspective of the outer apply, the function never returned an input as-is (since the inner apply actually applied alias).

Yeah, I think that is what is going on.